File: src/main/java/com/tobe/healthy/lessonhistory/domain/entity/LessonHistory.java
package com.tobe.healthy.lessonhistory.domain.entity;

import com.tobe.healthy.common.BaseTimeEntity;
import com.tobe.healthy.member.domain.entity.Member;
import com.tobe.healthy.schedule.domain.entity.Schedule;
import jakarta.persistence.*;
import lombok.AccessLevel;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.ToString;
import org.hibernate.annotations.DynamicUpdate;

import java.util.ArrayList;
import java.util.List;

import static jakarta.persistence.CascadeType.ALL;
import static jakarta.persistence.EnumType.STRING;
import static jakarta.persistence.FetchType.LAZY;
import static jakarta.persistence.GenerationType.IDENTITY;

@Entity
@DynamicUpdate
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@ToString
public class LessonHistory extends BaseTimeEntity<LessonHistory, Long> {

    @Id
    @GeneratedValue(strategy = IDENTITY)
    @Column(name = "lesson_history_id")
    private Long id;

    private String title;

    private String content;

    @OneToMany(fetch = LAZY, mappedBy = "lessonHistory", cascade = ALL)
    @ToString.Exclude
    private List<LessonHistoryComment> lessonHistoryComment = new ArrayList<>();

    @OneToMany(fetch = LAZY, mappedBy = "lessonHistory", cascade = ALL, orphanRemoval = true)
    @ToString.Exclude
    private List<LessonHistoryFiles> files = new ArrayList<>();

    @ManyToOne(fetch = LAZY)
    @JoinColumn(name = "trainer_id")
    @ToString.Exclude
    private Member trainer;

    @ManyToOne(fetch = LAZY)
    @JoinColumn(name = "student_id")
    @ToString.Exclude
    private Member student;

    @ManyToOne(fetch = LAZY)
    @JoinColumn(name = "schedule_id")
    @ToString.Exclude
    private Schedule schedule;

    @Enumerated(STRING)
    private LessonHistoryReadStatus feedbackChecked = LessonHistoryReadStatus.UNREAD;

    private LessonHistory(String title, String content, Member trainer, Member student, Schedule schedule) {
        this.title = title;
        this.content = content;
        this.trainer = trainer;
        this.student = student;
        this.schedule = schedule;
    }

    public void updateLessonHistory(String title, String content) {
        this.title = title;
        this.content = content;
    }

    public void updateFeedbackChecked() {
        this.feedbackChecked = LessonHistoryReadStatus.READ;
    }

    public static LessonHistory register(String title, String content, Member student, Member trainer, Schedule schedule) {
        return new LessonHistory(title, content, trainer, student, schedule);
    }
}

File: src/main/kotlin/com/tobe/healthy/lessonhistory/domain/entity/LessonHistory.kt
package com.tobe.healthy.lessonhistory.domain.entity

import com.tobe.healthy.common.BaseTimeEntity
import com.tobe.healthy.lessonhistory.domain.entity.LessonHistoryReadStatus.UNREAD
import com.tobe.healthy.member.domain.entity.Member
import com.tobe.healthy.schedule.domain.entity.Schedule
import jakarta.persistence.*
import jakarta.persistence.CascadeType.ALL
import jakarta.persistence.EnumType.STRING
import jakarta.persistence.FetchType.LAZY
import jakarta.persistence.GenerationType.IDENTITY
import lombok.ToString
import org.hibernate.annotations.DynamicUpdate

@Entity
@DynamicUpdate
@ToString
class LessonHistory(

    var title: String,

    var content: String,

    @OneToMany(fetch = LAZY, mappedBy = "lessonHistory", cascade = [ALL])
    @ToString.Exclude
    val lessonHistoryComment: MutableList<LessonHistoryComment> = mutableListOf(),

    @OneToMany(fetch = LAZY, mappedBy = "lessonHistory", cascade = [ALL], orphanRemoval = true)
    @ToString.Exclude
    var files: MutableList<LessonHistoryFiles> = mutableListOf(),

    @ManyToOne(fetch = LAZY)
    @JoinColumn(name = "trainer_id")
    @ToString.Exclude
    val trainer: Member? = null,

    @ManyToOne(fetch = LAZY)
    @JoinColumn(name = "student_id")
    @ToString.Exclude
    val student: Member? = null,

    @ManyToOne(fetch = LAZY)
    @JoinColumn(name = "schedule_id")
    @ToString.Exclude
    val schedule: Schedule? = null,

    @Enumerated(STRING)
    var feedbackChecked: LessonHistoryReadStatus = UNREAD,

    @Id
    @GeneratedValue(strategy = IDENTITY)
    @Column(name = "lesson_history_id")
    val id: Long? = null
) : BaseTimeEntity<LessonHistory, Long>() {

    fun updateLessonHistory(title: String, content: String) {
        this.title = title
        this.content = content
    }

    fun updateFeedbackStatus(feedbackStatus: LessonHistoryReadStatus) {
        this.feedbackChecked = feedbackStatus
    }

    companion object {
        fun register(title: String, content: String, student: Member, trainer: Member, schedule: Schedule): LessonHistory {
            return LessonHistory(
                title = title,
                content = content,
                trainer = trainer,
                student = student,
                schedule = schedule
            )
        }
    }
}
